experimental/delegate/linalg.pdl.mlir

```diff
@@ -0,0 +1,122 @@
+// PDL pattern spec to match an MLP and offload to an external function
+//
+// ```
+// void mlp_external(void *params, void *context, void *reserved)
+// ```
+//
+// which is the expected signature of an external function implemented
+// provided by a system plugin. See
+// samples/custom_dispatch/cpu/plugin/system_plugin.c for an example.
+//
+// The `params` is the following struct
+//
+// ```
+// struct mlp_params_t {
+//   const float *restrict lhs;
+//   size_t lhs_offset;
+//   const float *restrict rhs;
+//   size_t rhs_offset;
+//   int32_t M;
+//   int32_t N;
+//   int32_t K;
+//   float *restrict result;
+//   size_t result_offset;
+// };
+// ```
+//
+// In MLIR this corresponds to the function
+//
+// ```
+// func.func @mlp_external(%lhs : memref<..xf32>, %rhs : memref<..xf32>,
+//     %M: i32, %N : i32, %K : i32, %result : memref<..xf32>)
+// ```
+//
+// Note: In the above struct a `pointer, offset` pair represents a buffer
+// passed into the external function. So any access to `lhs`, `rhs` and
+// `result` is valid only if accessed as `lhs[lhs_offset + ...]`,
+// `rhs[rhs_offset + ]` and `result[result_offset + ...]`.
+pdl.pattern @mlp : benefit(1) {
+
+  // PDL matcher to match the MLP computation. This pattern is expected to
+  // match
+  //
+  // ```
+  // %result = func.call @mlp_external(%lhs : tensor<...xf32>,
+  //     %rhs : tensor<..xf32>, %M : i32, %N : i32, %K : i32) -> tensor<..xf32>
+  // ```
+  %lhs = pdl.operand
+  %rhs = pdl.operand
+  %lhs_type = pdl.type : tensor<?x?xf32>
+  %rhs_type = pdl.type : tensor<?x?xf32>
+  %matmul_type = pdl.type : tensor<?x?xf32>
+
+  %zero_val = pdl.attribute = 0 : index
+  %one_val = pdl.attribute = 1 : index
+  %zero_val_f32 = pdl.attribute = 0.000000e+00 : f32
+  %index_type = pdl.type : index
+  %f32_type = pdl.type : f32
+
+  %zero_op = pdl.operation "arith.constant" {"value" = %zero_val} -> (%index_type : !pdl.type)
+  %zero = pdl.result 0 of %zero_op
+  %one_op = pdl.operation "arith.constant" {"value" = %one_val} -> (%index_type : !pdl.type)
+  %one = pdl.result 0 of %one_op
+  %zero_f32_op = pdl.operation "arith.constant" {"value" = %zero_val_f32} -> (%f32_type : !pdl.type)
+  %zero_f32 = pdl.result 0 of %zero_f32_op
+
+  %m_op = pdl.operation "tensor.dim"(%lhs, %zero : !pdl.value, !pdl.value) -> (%index_type : !pdl.type)
+  %m = pdl.result 0 of %m_op
+  %n_op = pdl.operation "tensor.dim"(%rhs, %one : !pdl.value, !pdl.value) -> (%index_type : !pdl.type)
+  %n = pdl.result 0 of %n_op
+
+  %empty_op = pdl.operation "tensor.empty" (%m, %n : !pdl.value, !pdl.value) -> (%matmul_type : !pdl.type)
+  %empty = pdl.result 0 of %empty_op
+  
+  %fill_op = pdl.operation "linalg.fill" (%zero_f32, %empty : !pdl.value, !pdl.value) -> (%matmul_type : !pdl.type)
```

Comment:
I think you can avoid all the `tensor.dim` matching and just match the `linalg.fill` and `arith.constant 0.0 : f32` , capturing the `tensor.empty` as an operand.